tsl/src/continuous_aggs/create.c

```diff
@@ -298,6 +334,76 @@ cagg_create_hypertable(int32 hypertable_id, Oid mat_tbloid, const char *matpartc
 				 errmsg("could not create materialization hypertable")));
 }
 
+/*
+ * Add a logical decoding slot for reading invalidations.
+ *
+ * If a logical slot do not exist, it will be created. If it already exists,
+ * nothing will happen.
+ *
+ * This code is mostly copied from create_logical_replication_slot and friends
+ * in slotfunc.c.
+ *
+ * It is important that the slot exists if and only if there are continuous
+ * aggregates that use the WAL to collect invalidations.
+ *
+ * Obviously, the slot is needed to read invalidations for the hypertables and
+ * it need to have the version expected by the timescale extension. If the
+ * versions do not match, the returned rows will not match what is expected.
+ *
+ * Less obvious, the slot need to be removed once there are no more continuous
+ * aggregates that need the slot. If not, the WAL will not be used and will
+ * start to fill up storage.
+ *
+ * There is a bug in PostgreSQL causing it to error out if you have too many
+ * "old" transactions that need to be cleaned up, so running this when you
+ * have a lot of concurrent transactions might cause it to error out before
+ * the slot is created.
+ */
+static void
+cagg_add_logical_decoding_slot(void)
```

Comment:
When adding this replication slot, do we currently have anything using it? I am worried that if we add a slot, but have nothing that advances it, the WAL will be held back by the slot and it may grow very big. 

Perhaps we should not add the replication slot until we actually have something reading it.

Reply:
> When adding this replication slot, do we currently have anything using it? I am worried that if we add a slot, but have nothing that advances it, the WAL will be held back by the slot and it may grow very big.

If you define continuous aggregates, you either define a policy, have an external procedure to do the refresh, or refresh regularly. There is little point in defining one otherwise, so I think this is very unlikely.

> Perhaps we should not add the replication slot until we actually have something reading it.

Unfortunately that is not possible. The entire point of having a replication slot is to ensure that the WAL is not removed and that you keep a state of how much of the WAL you have seen.